### PR TITLE
[Feature:CourseMaterials] move files

### PR DESCRIPTION
**Pull request by 04vmatsibekker**
### Please check if the PR fulfills these requirements:

* [ ] Tests for the changes have been added/updated (if possible)
* [ ] Documentation has been updated/added if relevant

### What is the current behavior?

Instructors can upload file, but once they do so, there is no way to change file name/file location
### What is the new behavior?
Now instructors can change the file name/location, and display a different file name (link_title, for now) to non-owners of file.

New UI for both file/URL changes
Edit

### CODE REVIEW

**Comment by 04vmatsibekker** [comment]
#9464

**Comment by williamjallen** [comment]
Is there a reason why you made multiple PRs with the same changes and closed one of them?  This seems to be almost identical to #9464 and #9497.

**Review by ziesski — APPROVED** [review]
Looks good :)

File: site/app/controllers/MiscController.php [hunk 1]
```diff
@@ -271,7 +272,23 @@ public function downloadCourseFile($dir = null, $path = null, $course_material_i
             CourseMaterialsUtils::insertCourseMaterialAccess($this->core, $path);
         }
 
-        $filename = pathinfo($path, PATHINFO_BASENAME);
+        if ($dir == 'submissions') {
```
Comment (cjreed121) [inline site/app/controllers/MiscController.php:275]:
```suggestion
        if ($dir === 'submissions') {
```

File: site/app/controllers/MiscController.php [hunk 2]
```diff
@@ -271,7 +272,23 @@ public function downloadCourseFile($dir = null, $path = null, $course_material_i
...
+            //cannot download scanned images for bulk uploads
+            if (
+                strpos(basename($path), "upload_page_") !== false
+                && FileUtils::getContentType($path) !== "application/pdf"
+            ) {
+                $this->core->getOutput()->showError(self::GENERIC_NO_ACCESS_MSG);
+                return false;
+            }
```
Comment (cjreed121) [inline site/app/controllers/MiscController.php:283]: Is there a reason adding this is in this PR? I feel it should be a separate PR?

File: site/app/controllers/course/CourseMaterialsController.php [hunk 3]
```diff
@@ -438,30 +438,108 @@ public function ajaxEditCourseMaterialsFiles(bool $flush = true): JsonResponse {
...
+            $upload_path = FileUtils::joinPaths($this->core->getConfig()->getCoursePath(), "uploads", "course_materials");
+            $requested_path = $_POST['file_path'];
+            $new_path = FileUtils::joinPaths($upload_path, $requested_path);
+
+            $lastSlashPos = strrpos($new_path, "/");
+            if (isset($_POST['title'])) {
+                $file_name = $_POST['title'];
+                $new_path = substr($new_path, 0, $lastSlashPos + 1);
+                $new_path = FileUtils::joinPaths($new_path, $file_name);
+            }
```
Comment (cjreed121) [inline site/app/controllers/course/CourseMaterialsController.php:455]: You could use the php function `dirname` to clean this up a bit and not do as many substrings and position checks

File: site/app/libraries/FileUtils.php [hunk 4]
```diff
@@ -765,4 +765,32 @@ public static function getTopEmptyDir(string $path, string $base_path, &$results
             $results[] = $parent_dir;
         }
     }
+
+
+    /**
+     * Given a path, searches for the topmost directory that's empty, here empty means having no files but there can be
+     * empty subdirectories inside that directory
```
Comment (cjreed121) [inline site/app/libraries/FileUtils.php:772]: Update this comment

File: site/app/libraries/FileUtils.php [hunk 5]
```diff
@@ -765,4 +765,32 @@ public static function getTopEmptyDir(string $path, string $base_path, &$results
...
+            '<',    // Less than
+            '>',    // Greater than
+            '|',    // Pipe
+            '\0'    // Null character
+        ];
+        foreach ($disallowed as $char) {
+            if (strpos($path, $char) !== false) {
+                return false;
+            }
+        }
```
Comment (cjreed121) [inline site/app/libraries/FileUtils.php:793]: The file `test..test.out` is a valid filename which does not cause a parent directory issue. Are we now blocking that?

File: site/app/controllers/course/CourseMaterialsController.php [hunk 6]
```diff
@@ -438,30 +438,108 @@ public function ajaxEditCourseMaterialsFiles(bool $flush = true): JsonResponse {
...
-                $path = FileUtils::joinPaths($path, $file_name);
-                FileUtils::writeFile($path, "");
-                unlink($course_material->getPath());
-                $course_material->setUrlTitle($_POST['link_title']);
-                $course_material->setPath($path);
+
+
+
+                FileUtils::writeFile($new_path, "");
+                rename($course_material->getPath(), $new_path);
```
Comment (cjreed121) [inline site/app/controllers/course/CourseMaterialsController.php:533]: Why not only rename?

**Review by cjreed121 — CHANGES_REQUESTED** [review]

File: site/app/libraries/FileUtils.php [hunk 7]
```diff
@@ -767,4 +767,37 @@ public static function getTopEmptyDir(string $path, string $base_path, &$results
...
+
+        foreach ($disallowed as $char) {
+            if (strpos($path, $char) !== false) {
+                return false;
+            }
+        }
+
+        if (substr($path, 0, 2) === '..' || substr($path, -2) === '..') {
+            return false;
+        }
```
Comment (cjreed121) [inline site/app/libraries/FileUtils.php:799]: Is this just checking if the first or last 2 characters are `..`? A better check you can do is too iterate over all of the directory names and see if they equal `..`. You can just split by `/`.

**Review by cjreed121 — CHANGES_REQUESTED** [review]

**Review by NiaHeermance — CHANGES_REQUESTED** [review]
I am able to move files around and download them as intended.

**Review by NiaHeermance — APPROVED** [review]
I am able to move files to new paths as well. I approve!

**Pull request merged**